src/Symfony/Component/Encryption/Sodium/SodiumEncryption.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Encryption\Sodium;
+
+use Symfony\Component\Encryption\Ciphertext;
+use Symfony\Component\Encryption\EncryptionInterface;
+use Symfony\Component\Encryption\Exception\DecryptionException;
+use Symfony\Component\Encryption\Exception\EncryptionException;
+use Symfony\Component\Encryption\Exception\InvalidKeyException;
+use Symfony\Component\Encryption\Exception\SignatureVerificationRequiredException;
+use Symfony\Component\Encryption\Exception\UnableToVerifySignatureException;
+use Symfony\Component\Encryption\Exception\UnsupportedAlgorithmException;
+use Symfony\Component\Encryption\KeyInterface;
+
+/**
+ * Using the Sodium extension to safely encrypt your data.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ *
+ * @experimental in 6.0
+ */
+final class SodiumEncryption implements EncryptionInterface
+{
+    public function generateKey(string $secret = null): KeyInterface
```

Comment:
`$secret` is not a good name for this parameter. Maybe `$symmetric`?

Reply:
Hm.. Yes. A user may wrongly assume that if they generate two keys with the same secret, the keys would be identical. 

I dont like `$symmetric` because it is an adjective and it also leaks implementation details. 

Btw, the `SodiumKey` has a method `fromSecret()`.

Reply:
Symmetric isn't a leaked implementation detail. You fundamentally have to treat Asymmetric Encryption and Symmetric Encryption as separate things when you deal with encryption.

See https://paragonie.com/blog/2015/08/you-wouldnt-base64-a-password-cryptography-decoded for more.

Reply:
This is not encryption. This is an abstraction over encryption. 

As you can see, the implementation is very strict with separating symmetric and asymmetric encryption.

Reply:
Even still, `$secret` is a bad parameter name here.

Reply:
Let me elaborate:

If you're generating an asymmetric key pair, it consists of a secret key and a public key.
If you're generating a symmetric key, you only have that one key. 

Which means: That you're **still always generating a secret key**, regardless of that boolean's state.

"Symmetric" or "shared" are better names for the parameter, which also needs to percolate down to the interface and any classes that implement it.